### DISCUSSION TITLE
Proper use of HB

### DISCUSSION THREAD

**Discussion by kmihaylov** [Xyce/Xyce#16] [Q&A]
Hello. In the past I tried HB to simulate the impedance of a diode at RF frequencies. I remember that I had some unexpected node voltages and since I continue this simulation after a long time, now maybe I misunderstand some of the ways HB works.

Here is a little netlist that I simulate. I expect the voltage at node 1 to be the source voltage:
```
.global_param sourcevoltage=2

Vin 1 0 SIN(0 {sourcevoltage} 2e9)
R1 1 0 50

.options hbint numfreq=1

.hb 2e9
.print hb FORMAT=gnuplot vr(1)

.end
```

The output doesn't show 2V voltage at node 1:
```
Index       FREQ              VR(1)      
0       -2.00000000e+09    7.40148683e-17
1        0.00000000e+00   -1.48029737e-16
2        2.00000000e+09    7.40148683e-17
End of Xyce(TM) Simulation
```

I believe that I'm missing something here. Would you please give me some directions on how to properly use HB?

Thanks!

**Comment by eskiyerli** [comment]
You are not applying any dc bias. Of course you will not see any dc 
component.
I changed your netlist to the following netlist:

.global_param sourcevoltage=2

Vin 1 0 dc 1 SIN(1 {sourcevoltage} 2e9)
R1 1 0 50

.options hbint numfreq=1

.hb 2e9
.print hb FORMAT=gnuplot vr(1)

.endp

The output is now:
Index       FREQ              VR(1)
0       -2.00000000e+09    1.48029737e-16
1        0.00000000e+00    1.00000000e+00
2        2.00000000e+09    1.48029737e-16
End of Xyce(TM) Simulation

It now has 1V at DC as you did expect.

Regards,

Murat Eskiyerli

**Reply by kmihaylov** [reply]
Thank you for your answer,

As I'm aware of the DC level (0 in my case) I expect that vr(1) should be the peak or RMS voltage for 2GHz frequency. Should I use another probe or maybe I should modify the schematic?

Isn't your modification actually putting 1V DC offset? If so, I completely don't want this to be the case, since I expect the 2GHz AC source with the 50 Ohm series resistor to behave like 50 Ohm antenna.

I provided this oversimplified schematic to understand the basics of HB analysis. If I use transient analysis, for example, then I clearly can see the AC voltage component.

By the way, I use HB in order to solve large-signal problem (impedance of a diode w/o DC bias). Would Transient analysis still simulate large-signal scenario?

Thanks!
Kaloyan

**Comment by eskiyerli** [comment]
Sorry I misunderstood your question. Just try to replace vr(1) with 
vm(1) or add vi(1). The result is then:
Index       FREQ              VR(1)             VI(1)
0       -2.00000000e+09    1.11022302e-16    1.00000000e+00
1        0.00000000e+00   -2.22044605e-16    0.00000000e+00
2        2.00000000e+09    1.11022302e-16   -1.00000000e+00
End of Xyce(TM) Simulation

**Reply by kmihaylov** [reply]
Wow, it makes a lot more sense now! Maybe vr() for some reason doesn't fit in HB analysis...
I guess that the voltage value for the voltage source is the peak-to-peak value, but the vm(r1) value is the peak value (/2).

Here is my test script which works well now:
```

.global_param sourcevoltage=1

Vin 1 0 SIN(0 {sourcevoltage} 2e9)
R1 1 0 50

.options hbint numfreq=1

.hb 2e9
.print hb FORMAT=gnuplot vm(1) im(r1) {vm(1)/im(r1)}

.step sourcevoltage 0.01 1 0.1

.end
```
{VM(1)/IM(R1)} shows always 50, which is the value of the resistor. I'll upgrade the netlist with the diode and etc.

Would you please give me some directions on these questions:
1. Should I stick only with HB for large signal simulation, or Transient analysis will also calculate a large-signal problem?
2. Some manufacturers say that they offer small-signal models for their components. If I get the spice values for a diode, should I obey some specific large-signal model or it will be enough to use Is (saturation current), n (ideaity factor), Cj (junction capacitance) and etc?

Thank you very much!

**Comment by tvrusso** [comment]
You may not quite be getting what the output of HB really means.  These are, effectively, the complex coefficients of a Fourier expansion, and it is not simply the case that the real part is the magnitude as you're expecting.   In fact, in your case, the peak-to-peak magnitude is recovered by a combination of the imaginary parts of two lines of output.

Forgive me if I'm going into too much obvious detail, but perhaps the following will help explain.  The time-domain and frequency domain outputs are related by:
$$v(t)=\sum_{i=-N}^{i=N} v(f_i)e^{2\pi j f_i t}$$

Writing that out more explicitly:
$$v(t)=\sum_{i=-N}^{i=N} (Re(v(f_i) + j Im(v(f_i)))(\cos({2\pi f_i t}) + j\sin(2\pi f_i t))$$

or:
$$v(t) = \sum_{i=-N}^{i=N}(Re(v(f_i))\cos(2\pi f_i t) + jRe(v(f_i))\sin(2\pi f_i t) + jIm(v(f_i))\cos(2\pi f_i t) - Im(v(f_i))\sin(2\pi f_i t ))$$


Since your input is a sin and not a cos, the imaginary parts of the output are going to be the nonzero outputs, and the signal is divided up between the positive and negative frequencies.  In your case, since there's only output at the fundamental frequency, and the real parts are effectively zero as is the DC offset, you're getting:
$$v(t)=(-Im(v(-2))\sin(2\pi (-2GHz) t) - Im(v(2))\sin(2\pi (2GHz) t)$$

(the complex parts all cancel because the two nonzero components of V are complex conjugates and cosine is an even function).
or, plugging in the values from the output above and making use of the asymmetry of the sin function:
$$v(t)=1\sin(2\pi (2GHz) t) - (-1)\sin(2\pi (2GHz) t)$$
or:
$$v(t)=2\sin(2\pi (2GHz) t)$$
where the coefficient of the sin function is 2, as you expect.

To compute complex impedance, you need to divide the complex voltage by the complex current, not just the magnitudes (for a simple resistor this is a distinction without a difference, but when you start messing with nonlinear devices like diodes it will matter).  In older versions of Xyce this could not be done with the expression library directly and required explicit, real-equivalent arithmetic.   The new expression library in Xyce can handle complex arithmetic, but may require some tricks to print correctly.  @Karlsefni2012  may be able to help there.

**Reply by Karlsefni2012** [reply]
I haven't followed this whole thread (yet).  But to answer the question about complex arithmetic, if you want to output a complex voltage divided by a complex current, the following should work.  This is the `.print `line taken from the nelist, above, but with two more outputs added:

```
.print hb FORMAT=gnuplot vm(1) im(r1) {vm(1)/im(r1)} 
+ {re(v(1)/i(r1))} 
+ {img(v(1)/i(r1))} 
```

Note that if you just say `{v(1)/i(r1)}`  Xyce's expression library will perform complex arithmetic.  But by default it will only print the real part.  To see the real and imaginary part you have to explicitly ask for them, using the `Re()` and `Img()` operators.

This capability (complex arithmetic in expressions) is very new, so I'd be interested in hearing about any issues you encounter with it.

**Reply by kmihaylov** [reply]
@tvrusso 
> Forgive me if I'm going into too much obvious detail,

It is not obvious for me. I never paid attention to math and now I learn what I should have learned in the university.

> Since your input is a sin and not a cos, the imaginary parts of the output are going to be the nonzero outputs

I don't get it, but give me some time to learn it. Do I take the imaginary part as the real number? I mean, since the input is sin, do real and imaginary 'change places'?

> (the complex parts all cancel because the two nonzero components of V are complex conjugates and cosine is an even function)

Another homework.

Thank you for your time!

@Karlsefni2012,
I'm using Xyce 7.2 Should I upgrade to current? Because I get some unexpected calculations. Example provided:
```

.global_param sourcevoltage=1

Vin 1 0 SIN(0 {sourcevoltage} 2e9)
R1 1 0 50
*D1 input 0 diode1
.model diode1 d IS=50e-9 N=1.08 Cj0=0.18p RS=6

.options hbint numfreq=1
.hb 2e9

.print hb FORMAT=gnuplot v(1) i(r1) {img(v(1))} {img(i(r1))} {img(v(1))/img(i(r1))} {img(v(1)/i(r1))}
.end
```

Output:
```
Index       FREQ            Re(V(1))          Im(V(1))          Re(I(R1))         Im(I(R1))        {IMG(V(1))}      {IMG(I(R1))}    {IMG(V(1))/IMG(I(R1))} {IMG(V(1)/I(R1))}
0       -2.00000000e+09    3.70074342e-17    5.00000000e-01    1.15648232e-18    1.00000000e-02    5.00000000e-01    0.00000000e+00    1.00000000e+50    4.32345564e+17
1        0.00000000e+00   -7.40148683e-17    0.00000000e+00   -2.31296463e-18    0.00000000e+00    0.00000000e+00    0.00000000e+00   -1.00000000e+50   -0.00000000e+00
2        2.00000000e+09    3.70074342e-17   -5.00000000e-01    1.15648232e-18   -1.00000000e-02   -5.00000000e-01    0.00000000e+00   -1.00000000e+50   -4.32345564e+17
End of Xyce(TM) Simulation
```

v(1) or i(r1) both print the expected voltage and current. img(v1) also prints the expected voltage, but img(i(r1)) prints 0. 

What is the expected output of  img(v(1)/i(r1)) or img(v(1))/img(i(r1)). Do I misunderstand complex number calculation here?

Thanks!

**Comment by tvrusso** [comment]
And to answer one of your direct questions explicitly:  small signal analysis (looking at small signals around an operating point) is only performed in AC analysis.  Transient analysis and HB analysis do not use a small-signal model and would both be usable.  You could do a long transient analysis and then use an FFT on the late-time, periodic steady state you achieve and you should get the same Fourier coefficients that HB would give you.
